### DISCUSSION TITLE
Data Hub | Performance with big data sets

### DISCUSSION THREAD

**Discussion by smitterer** [pimcore/pimcore#13742] [Q&A / Support]
Hi,

we are going to use the PIM part of Pimcore to store our approx. 600.000 products and its approx. 3.000 attributes (while most products only have 10 - 30 attributes).

To get the product information to other systems like odoo, we'd like to use Data Hub and convert the received data to SQL / plpgsql and directly import into the remote DB.

Does anyone have some experience in requesting such a huge number of products and their attributes via GraphQL?
How should such a query look like and can we set Data Hub to return all available attributes of a product?

Besides the initial export of all data, we'd like to only make available those products (via a separate GraphQL) which where updated. Is it possible to set something like a flag when an attribute of a product changed and only provide those products via this separate GraphQL? Or would GraphQL subscriptions work here?
After the remote system received the data, it should de-flag the product as "to be synced". Any idea on how this could be accomplished?

If Data Hub is not optimal for this, we might also build a PHP API to pimcore, while questions on the number of products and so on remain.

Thanks!

**Comment by SamyMP** [comment]
Are you familiar with GraphQL ? If not, you can see how a query looks like in the [official documentation](https://graphql.org/learn/queries/)

It is possible to return several products thanks to the listing interfaces provided by DataHub. You can prevent timeouts by paginating thanks to the `first` and `after` parameters.

You can also pass filters to the query, and specify that you want to return objects that has been modified after a certain date so that you won't have to import everything.

Basically, for your initial import, you'll have a query that looks like this : 
```graphql
query getAllProducts($offset: Int) {
  getProductListing(first: 500, after: $offset) {
    edges {
      node {
        attribute1
        attribute2
        # So on...
      }
    }
  }
}
```

With the variable declaration : 
```graphql
{
  "offset": 0
}
```

This query retrieve the first 500 products of your database. Then, you just have to increment the offset by 500 in this exemple, to get the rest until you reach your total number of Product

The filter to use to get only modified data is the following : 

```graphql
# Replace '123456789' by a timestamp corresponding to the date you want
filter: "{\"o_modificationDate\":{\"$gte\": 123456789}}"
```

Some useful links : 
https://graphql.org/learn/queries/#variables
https://pimcore.com/docs/data-hub/current/
https://pimcore.com/docs/data-hub/current/GraphQL/Query/Filtering.html
https://pimcore.com/docs/pimcore/6.9/Development_Documentation/Web_Services/Query_Filters.html

I hope this helps 🙂
